docs/src/adr/007adr-restructuring.md

```diff
@@ -0,0 +1,161 @@
+# ADR-007: Apalache Package Structure Guidelines
+
+| author       | revision |
+| ------------ | --------:|
+| Jure Kukovec |        1 |
+
+This ADR documents the design policies guiding the package and dependency structure of Apalache.
+When introducing new classes, use the guidelines defined below to determine which package to place them in.
+
+## 1. Level Structure
+
+We define Apalache architecture in terms of enumerated levels (L0, L1, L2, etc.). Each level may only hold dependencies belonging to a lower level.
+
+Levels are split into two categories:
+
+  1. Interface core: Classes within the interface core relate to general TLA+ concepts, and are intended to be usable by 3rd party developers.
+  2. Apalalache core: Classes in the Apalache core relate to specific Apalache- or model-checking- functionality
+
+## 2. Interface core
+
+Together, L0-L6 make up the interface core. Notably, these levels are Apalache-agnostic. The concepts of these levels are packaged as follows:
```

Comment:
What does it mean to "package the concepts" of these levels?

Reply:
What is the difference between a "level" and a package?Are the "levels" just a conceptual framing? Do they correspond to anything in the structure of the code or packaging?

Reply:
The way I originally approached this, levels were the linearization of the dependency graph, which should be a DAG and thus admit a topological sort. They don't really correspond to anything at the code level.